_P221_LandisGyrUH550.ino

```diff
@@ -0,0 +1,450 @@
+#define USES_P221
+#ifdef USES_P221
+//#######################################################################################################
+//############################# Plugin 218: read Landis & Gyr UH550 #####################################
+//#######################################################################################################
+
+//IR RX/TX sensor based on https://wiki.volkszaehler.org/hardware/controllers/ir-schreib-lesekopf
+
+//Device pin 1 = RX
+//Device pin 2 = TX
+
+#include <ESPeasySoftwareSerial.h>
+#define PLUGIN_221
+#define PLUGIN_ID_221 221
+#define PLUGIN_NAME_221 "Communication - Landis & Gyr UH550 [TESTING]"
+#define PLUGIN_VALUENAME1_221 "Heat"
+#define PLUGIN_221_VARIABLECOUNT 1
+
+// Search filters in received data.
+// NOTE: This data remains in Flash!
+const char initFilter[] PROGMEM = {"/LUGCUH50"};
+const char qh_Filter[] PROGMEM = {"6.8("};                // 6.8 - Quantity of Heat (GJ)
+const char vol_Filter[] PROGMEM = {"6.26("};              // 6.26 - Volume (m3)
+const char qhPrev_Filter[] PROGMEM = {"6.8*01("};         // 6.8*01 - Quantity of Heat previous year (GJ)
+const char volPrev_Filter[] PROGMEM = {"6.26*01("};       // 6.26*01 - Volume previous year (m3)
+const char poMax_Filter[] PROGMEM = {"6.6("};             // 6.6 - Power Max (kW)
+const char inOutFlowMaxTemp_Filter[] PROGMEM = {"9.4("};  // 9.4 - max inflow & max reflow temperature (C°)
+
+const char* const filters[] PROGMEM = { qh_Filter, vol_Filter, inOutFlowMaxTemp_Filter, qhPrev_Filter, volPrev_Filter, poMax_Filter }; 
+const byte numberFilters = sizeof(filters) / sizeof(filters[0]);
+
+boolean Plugin_221_init = false;
+byte PIN_SER_RX = 0;
+byte PIN_SER_TX = 0;
+
+struct ValueUnitPair
+{
+  char value[10];
+  char unit[10];
+};
+
+struct DataSet
+{
+  char address[10];
+  byte vupCount;
+  struct ValueUnitPair vup[20];
+};
+
+struct DataSet datasets[10];

```

Comment:
That's a lot of allocated memory on the stack.
`ValueUnitPair` takes 20 bytes
`DataSet` takes 20 * 20 bytes + 11 bytes (maybe more due to alignment)
So that's 411 bytes.

And then you allocate 10 or them, which makes it 4110 bytes.

It is allocated not on cont stack (Arduino space) but on the sys stack (core library space)
But recently the cont stack was moved 4 kbyte down to partly overlay the sys stack, so I'm not sure if there is 4k left on there.

Has this been tested with recent core libraries? I doubt it.

Reply:
ESP82xx Core 2_4_2, NONOS SDK 2.2.1(cfd48f3), LWIP: 2.0.3

with above libs I ran it stable on on a wemos. I allocated static to avoid memory leaks and fragmentation. Unfortunately the smartmeter has lots of values that have to be parsed.